File: Good-First-Issues/palindrome_linked_list.py
class ListNode:
   def __init__(self, data, next = None):
      self.data = data
      self.next = next

def make_list(elements):
   head = ListNode(elements[0])
   for element in elements[1:]:
      ptr = head
      while ptr.next:
         ptr = ptr.next
      ptr.next = ListNode(element)
   return head
class Solution(object):
   def isPalindrome(self, head):
      fast,slow = head,head
      rev = None
      flag = 1
      if not head:
         return True
      while fast and fast.next:
         if not fast.next.next:
            flag = 0
            break
         fast = fast.next.next
         temp = slow
         slow = slow.next
         temp.next = rev
         rev = temp
      #print(fast.val)
      fast = slow.next
      slow.next = rev
      if flag:
         slow = slow.next
      while fast and slow:
         if fast.data != slow.data:
            return False
         fast = fast.next
         slow = slow.next
      return True

t=int(input())
for i in range (t) :
    l=list(map(int,input().split()))
    if -1 in l:
        l.pop(-1)
    elif len(l)==1:
        print(True)
    head = make_list(l)
    ob1 = Solution()
    print(ob1.isPalindrome(head))

File: Good-First-Issues/Palindrome_LinkedList.cpp
#include <bits/stdc++.h>
using namespace std;


struct LL{
  int data;
  LL *next;  
};


void insertAtBeginning(struct LL**head,int dataToBeInserted)
{
  struct LL* curr=new LL;
    
    curr->data=dataToBeInserted;
    curr->next=NULL;
  
  if(*head==NULL)
      *head=curr; 
  else
    {
      curr->next=*head; 
      *head=curr;
    }
}
void reverse(struct LL **head)
{
  struct LL *temp=NULL,*prev=NULL,*curr=*head;
  while(curr!=NULL)
  {
    temp=curr->next;
  curr->next=prev;
  prev=curr;
  curr=temp;
  }
  *head=prev;
}
struct LL* middleOfList(struct LL**head)
{
  
  struct LL*slow=*head,*fast=*head;
  while(fast&&fast->next)
    {
      slow=slow->next;
      fast=fast->next->next;
    }
  return  slow;
  
}


bool isPalindrome(struct LL *head, struct LL *middle)
{
  
  while(middle != NULL)
  {
    if(middle->data != head->data)
      return false;
    head = head->next;
    middle = middle->next;
  }
  
  return true;
}

int main()
{
    int t;
    cin>>t;

    while(t--){
  
        struct LL *head = NULL;
        
        int n;
        cin>>n;

        int arr[n];
        for(int i=0; i<n; i++){
            cin>>arr[i];
        }

        for(int i=0; i<n; i++){
          insertAtBeginning(&head,arr[i]);
        }
  
        struct LL * middle = middleOfList(&head);
        reverse(&middle);
  
        if(isPalindrome(head,middle))
        cout<<"false\n";
  
        else
        cout<<"true \n";
    }
  
  return 0;
}